add(chain): Impl `zcash_protocol::consensus::Parameters` for `Network`

upbqdn opened the pull request:
## Motivation

The new scanning API in `zcash_client_backend` requires the `zcash_protocol::consensus::Parameters` trait for the passed network. This is a preparatory PR for the transition to the new API.

### PR Author Checklist

#### Check before marking the PR as ready for review:
  - [x] Will the PR name make sense to users?
  - [x] Does the PR have a priority label?
  - [ ] Have you added or updated tests?
  - [x] Is the documentation up to date?

## Solution

- [Use full paths for types from zcash_primitives](https://github.com/ZcashFoundation/zebra/commit/09920d045eb55252c8532b2f71dcebab39fd7778)
- [Impl and use conversion between network kinds](https://github.com/ZcashFoundation/zebra/commit/9a45ada2a8fbb8e0103ab49b455549588ead6986)
- [Impl conversions between network upgrades](https://github.com/ZcashFoundation/zebra/commit/3084f3bb697f2df7f18fbc92ad4e61c2a5daec5b)
- [Impl zcash_protocol::consensus::Parameters](https://github.com/ZcashFoundation/zebra/commit/ba6c0d7110925609c61cd27e61c5f2d6a72a7a19)

### Reviewer Checklist

Check before approving the PR:
  - [ ] Does the PR scope match the ticket?
  - [ ] Are there enough tests to make sure it works? Do the tests cover the PR motivation?
  - [ ] Are all the PR blockers dealt with?
        PR blockers can be dealt with in new tickets or PRs.

_And check the PR Author checklist is complete._

upbqdn committed:
Use full paths for types from `zcash_primitives`

Upcoming commits will use `zcash_protocol::consensus::Parameters`, which
would make names such as `zp_consensus::Parameters` ambiguous since such
a name could stand for both `zcash_primitives...` and
`zcash_protocol...`.

upbqdn committed:
Impl and use conversion between network kinds

These conversions will make implementing
`zcash_protocol::consensus::Parameters` for `Network` easier.

upbqdn committed:
Impl conversions between network upgrades

upbqdn committed:
Impl `zcash_protocol::consensus::Parameters`

upbqdn commented on zebra-chain/src/parameters/network_upgrade.rs:
```diff
@@ -493,6 +493,32 @@ impl NetworkUpgrade {
     }
 }
 
+impl From<zcash_protocol::consensus::NetworkUpgrade> for NetworkUpgrade {
+    fn from(nu: zcash_protocol::consensus::NetworkUpgrade) -> Self {
+        match nu {
+            zcash_protocol::consensus::NetworkUpgrade::Overwinter => Self::Overwinter,
+            zcash_protocol::consensus::NetworkUpgrade::Sapling => Self::Sapling,
+            zcash_protocol::consensus::NetworkUpgrade::Blossom => Self::Blossom,
+            zcash_protocol::consensus::NetworkUpgrade::Heartwood => Self::Heartwood,
+            zcash_protocol::consensus::NetworkUpgrade::Canopy => Self::Canopy,
+            zcash_protocol::consensus::NetworkUpgrade::Nu5 => Self::Nu5,
+        }
+    }
+}
+
+impl From<zcash_primitives::consensus::NetworkUpgrade> for NetworkUpgrade {
+    fn from(value: zcash_primitives::consensus::NetworkUpgrade) -> Self {
+        match value {
+            zcash_primitives::consensus::NetworkUpgrade::Overwinter => NetworkUpgrade::Overwinter,
+            zcash_primitives::consensus::NetworkUpgrade::Sapling => NetworkUpgrade::Sapling,
+            zcash_primitives::consensus::NetworkUpgrade::Blossom => NetworkUpgrade::Blossom,
+            zcash_primitives::consensus::NetworkUpgrade::Heartwood => NetworkUpgrade::Heartwood,
+            zcash_primitives::consensus::NetworkUpgrade::Canopy => NetworkUpgrade::Canopy,
+            zcash_primitives::consensus::NetworkUpgrade::Nu5 => NetworkUpgrade::Nu5,
+        }
+    }
+}
```
We might consider using one of the upstream `NetworkUpgrade` types instead of our own, and similarly for some other types.

arya2 replied:
That's a good idea, we could add `Genesis`/`BeforeOverwinter` variants to the upstream `NetworkUpgrade` type (or refactor our code to avoid using them) and implement existing methods on `NetworkUpgrade` as trait methods.

upbqdn committed:
Move `mod address` from under the gbt feature

upbqdn committed:
Merge branch 'main' into impl-Params-from-zcash-protocol

arya2 commented on zebra-chain/src/parameters/network.rs:
```diff
@@ -315,21 +308,25 @@ impl zp_consensus::Parameters for Network {
 
     fn hrp_sapling_extended_full_viewing_key(&self) -> &str {
         match self {
-            Network::Mainnet => zp_constants::mainnet::HRP_SAPLING_EXTENDED_FULL_VIEWING_KEY,
+            Network::Mainnet => {
+                zcash_primitives::constants::mainnet::HRP_SAPLING_EXTENDED_FULL_VIEWING_KEY
+            }
             Network::Testnet(params) => params.hrp_sapling_extended_full_viewing_key(),
         }
     }
 
     fn hrp_sapling_payment_address(&self) -> &str {
         match self {
-            Network::Mainnet => zp_constants::mainnet::HRP_SAPLING_PAYMENT_ADDRESS,
+            Network::Mainnet => zcash_primitives::constants::mainnet::HRP_SAPLING_PAYMENT_ADDRESS,
```
Optional/Nitpick: Could we skip these changes to avoid merge conflicts?

upbqdn replied:
If we skip these changes, we will have `zp_constants` referring to `zcash_primitives::contants`, and we'll also have `zcash_protocol::*`, which also has `constants`, although we only use `zcash_protocol::consensus` now. I tried to summarize the rationale in the commit msg. I'm happy to revert the changes, though.

upbqdn replied:
I meant the msg in this commit: [09920d0](https://github.com/ZcashFoundation/zebra/pull/8425/commits/09920d045eb55252c8532b2f71dcebab39fd7778).

arya2 replied:
Ah, got it, okay.

arya2 commented on zebra-chain/src/primitives/address.rs:
```diff
@@ -163,34 +163,42 @@ impl Address {
+    fn from(network: &zcash_address::Network) -> Self {
+        network.into()
+    }
+}
```
Nitpick: This seems like it might not be necessary?

upbqdn replied:
Fixed in [0aa9c69](https://github.com/ZcashFoundation/zebra/pull/8425/commits/0aa9c69298e4b1bf98972533516b3517a28419f4).

arya2 reviewed (dismissed):
This looks good to me, I'm not sure why the tests were hanging, maybe it was a transient issue? It may have been an issue with the DNS seeders, they're up but are only returning 1-2 initial peers.

upbqdn committed:
Remove a redundant `impl From`

upbqdn committed:
Merge branch 'main' into impl-Params-from-zcash-protocol

upbqdn committed:
Merge branch 'main' into impl-Params-from-zcash-protocol

upbqdn committed:
Impl `From` for a reference type correctly

The previous implementation called `into` on a reference without
dereferencing it. I thought the compiler would implicitly dereference,
but it didn't, which lead to infinite recursion since the fn ended up
calling itself.

upbqdn commented:
> I'm not sure why the tests were hanging

Resolved in [3a66152](https://github.com/ZcashFoundation/zebra/pull/8425/commits/3a66152d9e21c980ecace1a1ce55ea37bfd78dc9).

upbqdn committed:
Merge branch 'main' into impl-Params-from-zcash-protocol

arya2 commented:
@Mergifyio refresh

upbqdn committed:
Merge branch 'main' into impl-Params-from-zcash-protocol
